src/com/moclojer/components/logs.clj

```diff
@@ -1,62 +1,137 @@
+  map `m` to a string.
+
+  This is because OpenSearch only accepts string json values."
+  [v]
+  (cond
+    (map? v) (reduce-kv (fn [m k v] (assoc m k (->str-values v))) {} v)
+    (coll? v) (into (empty v) (map ->str-values v))
+    (string? v) v
+    (nil? v) v
+    :else (str v)))
+
+(defn signal->opensearch-log
+  "Adapts a telemere signal to a pre-defined schema for OpenSearch."
+  [{:keys [thread location] :as signal}]
+  (-> (select-keys signal [:level :ctx :data :msg_ :uid :inst])
+      (merge {"thread/group" (:group thread)
+              "thread/name" (:name thread)
+              "thread/id" (:id thread)
+              "location" (str (:ns location) ":"
+                              (:line location) "x"
+                              (:column location))})
+      (->str-values)
+      (assoc :error (:error signal))
+      (json/write-str)))
+
+(defn build-opensearch-base-req
+  [config]
+  (let [{:keys [username password host port index]} config
+        url (str "https://" host ":" port "/_bulk")]
+    {:method :post
+     :url url
+     :basic-auth [username password]
+     :content-type :json
+     :body (json/write-str {:index {:_index index}})}))
+
+(defn send-opensearch-log-req
+  [base-req log]
+  (try
+    (http-client/request
+     (update base-req :body str \newline log \newline))
+    (catch Exception e
+      (log-console! :error "failed to send opensearch log"
+                    {:log log} e))))
+
+;; If on dev `env`, does basically nothing besides setting the min
+;; level. On `prod` env however, an async channel waits for log events,
+;; which are then sent to OpenSearch.
+(defrecord Logger [config]
+  component/Lifecycle
+  (start [this]
+    (let [prod? (= (:env config) :prod)
+          os-cfg (:opensearch config)
+          log-ch (async/chan)
+          os-base-req (build-opensearch-base-req os-cfg)]
+
+      (t/set-min-level!
+       (or (get-in config [:logger :min-level]) :info))
+
+      (when (and prod? (instance? ManyToManyChannel log-ch))
+        (t/set-ns-filter! {:disallow #{"*jetty*" "*hikari*"
+                                       "*pedestal*" "*migratus*"}})
+
+        (t/add-handler!
+         ::opensearch
+         (fn [signal]
+           (async/go
+             (async/>! log-ch (signal->opensearch-log signal)))))
+
+        (async/go
+          (while true
+            (let [[log _] (async/alts! [log-ch])]
+              (send-opensearch-log-req os-base-req log)))))
+
+      (assoc this :log-ch log-ch)))
+  (stop [this]
+    (t/remove-handler! ::opensearch)
+    (update this :log-ch #(when % (async/close! %)))))
+
+(defn log
+  ([level msg]
+   (log level msg nil nil nil))
+  ([level msg data]
+   (log level msg data nil nil))
+  ([level msg data ctx]
+   (log level msg data ctx nil))
+  ([level msg data ctx error]
+   (t/log! {:level level
+            :ctx ctx
+            :data data
+            :error error}
+           (str msg))))
 
 (defn gen-ctx-with-cid []
```

Comment:
what does this ctx and cid stands for?

Reply:
before this API change, context referred to anything in a log that could identify it, besides the level and the message itself. Now, since we're changing from plain timbre to telemere, context has a whole new different meaning. It's the context that identifies a log in a series of other logs. There are other identifiable keys in the log, like :data and :error.

So, answering your question, :ctx used to mean any data that could follow the log. Now, :ctx is the specifc identification of a log in a series of logs, like the request where the log comes from. CID stands for context id, it's a random uuid generated on starting points, like requests and message queue job starts, so we can identify where the log is coming from.

Reply:
thx :pray: